Added radio buttons to select histogram method for surface fields

BernardClaxton opened the pull request:
<!-- Thanks for contributing! Please add a short description of your change, and link to an issue, e.g. "Fixes #123" -->

Fixes #1173

### Contribution checklist

Aim to have all relevant checks ticked off before merging. See the [developer's guide](https://metoffice.github.io/CSET/contributing/) for more detail.

- [x] Documentation has been updated to reflect change.
- [x] New code has tests, and affected old tests have been updated.
- [x] All tests and CI checks pass.
- [x] Ensured the pull request title is descriptive.
- [ ] Conda lock files have been updated if dependencies have changed.
- [ ] Attributed any Generative AI, such as GitHub Copilot, used in this PR.
- [x] Marked the PR as ready to review.

BernardClaxton commented:
Fixes #1173

jfrost-mo reviewed (changes requested):
The overall design looks very sensible and it should provide a welcome improvement in our histogram plots.

I've left a few comments throughout, but the two big things are adding tests for the histogram function, and considering whether we are right in making this user configurable outside of the recipes. See [this comment](https://github.com/MetOffice/CSET/pull/1303#discussion_r2031057196).

BernardClaxton commented on cset-workflow/meta/diagnostics/rose-meta.conf:
For histograms constructed using a single bin width, the difference between the frequency, normalised frequency an density histograms is the scaling on the y-axis. The morphology of the plots remains the same. So in this case the choice of plot is really down to a user's personal preference.

If a histogram were to be constructed from uneven bin widths, then differences between the plot would become apparent. But we are a long away off implementing this facility.

It did cross my mind that the parameters controlling the histogram plots ought to be in a JSON file, similar to the control of the colorbars. Then the radio buttons could define the default behaviour, with tweaks supplied by a JASON file.

BernardClaxton commented on cset-workflow/meta/diagnostics/rose-meta.conf:
```diff
@@ -138,10 +138,33 @@ type=python_boolean
+compulsory=true
+sort-key=0surface5
+
+[template variables=HISTOGRAM_METHOD_SURFACE]
```
I thought about this at the time. I couldn't decide whether we would require different histograms for surface and pressure plots, so I coded so as to allow this option for the future. If not required, then the HISTOGRAM_METHOD_SURFACE variable could be used for pressure plots.

jfrost-mo replied:
Given your reply here (https://github.com/MetOffice/CSET/pull/1303#discussion_r2031241831) I'd be tempted to go for a single setting.

You'll want to move this into the Setup category, and rename to remote the _SURFACE suffix.

jfrost-mo replied:
I've done this myself as part of updating this branch with the latest changes, as I had to go into the rose-meta.conf anyway.

BernardClaxton commented on cset-workflow/opt/rose-suite-RAL3LFRIC.conf:
Done

BernardClaxton commented on src/CSET/_workflow_utils/run_cset_recipe.py:
```diff
@@ -109,7 +109,11 @@ def run_recipe_steps():
+    if histogram_method_surface:
+        command.append(f"--histogram-method-surface={histogram_method_surface}")
     logging.info("Running %s", shlex.join(command))
+
```
Thanks for updating this.

BernardClaxton commented on src/CSET/operators/plot.py:
```diff
@@ -665,7 +670,7 @@ def _plot_and_save_scatter_plot(
     # plot the cube_x and cube_y 1D fields as a scatter plot. If they are CubeLists this ensures
     # to pair each cube from cube_x with the corresponding cube from cube_y, allowing to iterate
-    # over the pairs simultaneously.
+    # over the pairs simultaneously.src/CSET/operators/plot.py:271
```
Thanks, I've fixed this pasted typo.

BernardClaxton commented on src/CSET/operators/plot.py:
```diff
@@ -704,6 +709,80 @@ def _plot_and_save_scatter_plot(
+
+    frequency           - This is constructed from the raw counts within each bin.
+
+    normalised_freqency - The frequency counts are normailsed so that the total
```
The two typos `normalised_freqency` and `normailsed` have been corrected.

BernardClaxton commented on src/CSET/operators/plot.py:
```diff
@@ -704,6 +709,80 @@ def _plot_and_save_scatter_plot(
+    normalised_freqency - The frequency counts are normailsed so that the total
+                          counts is 1.
+
+    density             - The counts per bin are normalised to by both the total
```
The extraneous `to` has been deleted.

BernardClaxton commented on src/CSET/operators/plot.py:
```diff
@@ -704,6 +709,80 @@ def _plot_and_save_scatter_plot(
+    histogram and the number of bins to use. Using these options will set the bin
+    width to ( value_maximum - value_minimum ) / number of bins. See the documentation
+    for matplotlib.pyplot.hist for more details at:
+    https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.hist.html
```
The link has been updated as suggested.

BernardClaxton commented on src/CSET/operators/plot.py:
```diff
@@ -704,6 +709,80 @@ def _plot_and_save_scatter_plot(
+    ----------
+    data: array
+        Input values of the data to use for computing the histogram.
+    method: "frequency" | "normalised_freqency" | "density", optional
```
Another `freqency` typo corrected. I've done a `grep` to sniff out any more of these in `plot.py`, but this seems to be the last.

jfrost-mo commented:
Sorry for being slow to review this. I'll go ahead and update the branch myself to fix the conflicting changes.

BernardClaxton commented:
I've also been a bit busy James. I've tried to implement your suggested edits, but I've yet to write a test for the new histogram plotting function. I can also follow up on your suggestion to move the histogram method to the Setup.

jfrost-mo reviewed (commented):
Its looking good with the changes you have done. I've just brought the branch up-to-date with the trunk, so once we have a few more tests we will be ready to merge.

daflack committed:
Add radio buttons to select histogram method for surface fields

This is a squash commit that was everything on the branch

daflack reviewed (commented):
Looks good and works well. A really valuable change to CSET. 

One very minor comment worth thinking about (also partly not sure if it was something I changed in the re-base that has altered it).

I was wondering if it was worth putting a little bit of information in the help text to indicate that the method selection appears in `General Setup Options` or seeing if there is a way it can be moved to appear in the location where the histograms are selected.

Also worth noting there should probably be a separate pull result to allow pressure levels and model level histograms to be selected in the same way as it currently does not trigger for those).

BernardClaxton commented:
Originally the histogram method was selected where the histogram options appear. So yes, it might be worth popping this function back there.

This enhancement was designed to allow someone to select a single method for producing histograms. It might be worth considering allowing multiple methods to be selected, similar to how "MEAN", "MIN" and "MAX" can be selected for producing aggregated plots.

jfrost-mo committed:
Normalise casing of histogram mode

jfrost-mo committed:
Don't add CLI option for histogram method

Instead we will set it through a recipe variable.

jfrost-mo committed:
Add histogram_method argument to plot.plot_histogram_series

It is passed through to the internal plotting functions, but not all of
them use it yet.

jfrost-mo committed:
Pass histogram method through recipes
